server/metrics.go

```diff
@@ -87,59 +68,37 @@ func newStatistics() *statistics {
 	}
 }
 
-// logInvalid increments the error count and updates the error percentage.
+func (stats *statistics) logRequest(opcode protocol.Op) {
+	stats.requests.WithLabelValues(opcode.String()).Inc()
+}
+
 func (stats *statistics) logInvalid(opcode protocol.Op) {
-	stats.requestsInvalid.Inc()
+	stats.requestsInvalid.WithLabelValues(opcode.String()).Inc()
 }
 
-// logConnFailure increments the error count of connFailures.
 func (stats *statistics) logConnFailure() {
 	stats.connFailures.Inc()
 }
 
 func (stats *statistics) logKeyLoadDuration(loadBegin time.Time) {
-	stats.keyLoadDuration.Observe(float64(time.Now().Sub(loadBegin)) / float64(time.Second))
+	stats.keyLoadDuration.Observe(time.Since(loadBegin).Seconds())
 }
 
 // logRequestExecDuration logs the time taken to execute an operation (not
-// including queueing). If the operation is an RSA operation, primes is the
-// number of primes in the RSA private key; otherwise, it is ignored.
-func (stats *statistics) logRequestExecDuration(opcode protocol.Op, primes int, requestBegin time.Time) {
-	stats.requestExecDuration.Observe(float64(time.Now().Sub(requestBegin)) / float64(time.Second))
-	primesStr := ""
-	switch opcode {
-	case protocol.OpRSADecrypt, protocol.OpRSASignMD5SHA1, protocol.OpRSASignSHA1,
-		protocol.OpRSASignSHA224, protocol.OpRSASignSHA256, protocol.OpRSASignSHA384,
-		protocol.OpRSASignSHA512, protocol.OpRSAPSSSignSHA256,
-		protocol.OpRSAPSSSignSHA384, protocol.OpRSAPSSSignSHA512:
-		if primes > 10 {
-			// if a key has more than 10 primes, it is:
-			// a) degenerate, and so we don't care about measuring its performance
-			// b) probably a DoS attack to try to get us to generate tons of labels
-			primes = 0
-		}
-		primesStr = fmt.Sprint(primes)
-	}
-	stats.requestExecDurationByOpcode.WithLabelValues(opcode.String(), primesStr).
-		Observe(float64(time.Now().Sub(requestBegin)) / float64(time.Second))
+// including queueing).
+func (stats *statistics) logRequestExecDuration(opcode protocol.Op, requestBegin time.Time, err protocol.Error) {
+	stats.requestExecDuration.WithLabelValues(opcode.String(), err.String()).Observe(time.Since(requestBegin).Seconds())
```

Comment:
I would just have a "success", 1 for err == nil, 0 for err != nil, or something. err.String() has the potential to create a ton of tags and blow up opentsdb. I also don't like ErrNone.

Reply:
Ah, I thought `err` was of `type error` but it's of `type protocol.Error`. You're right, it's bounded by the size of the enum, and ErrNone is the type's zero value.